File: skywalker/network_utils.cpp
// Copyright (C) 2024 Michel de Boer
// License: GPLv3
#include "network_utils.h"
#include <QDebug>
#include <QtGlobal>

#ifdef Q_OS_ANDROID
#include <QJniObject>
#include <QOperatingSystemVersion>
#endif

namespace Skywalker::NetworkUtils {

int getBandwidthKbps()
{
#if defined(Q_OS_ANDROID)
    const int kbps = QJniObject::callStaticMethod<jint>(
        "com/gmail/mfnboer/NetworkUtils",
        "getBandwidthKbps", "()I");

    qDebug() << "Bandwidth:" << kbps << "kbps";
    return kbps;
#else
    return 10'000; // Use 10 mbps as default for Linux
#endif
}

bool isUnmetered()
{
#if defined(Q_OS_ANDROID)
    const bool unmetered = (bool)QJniObject::callStaticMethod<jboolean>(
        "com/gmail/mfnboer/NetworkUtils",
        "isUnmetered", "()Z");

    qDebug() << "Unmetered:" << unmetered;
    return unmetered;
#else
    return true;
#endif
}

}

File: android/src/com/gmail/mfnboer/NetworkUtils.java
// Copyright (C) 2023 Michel de Boer
// License: GPLv3

package com.gmail.mfnboer;

import com.gmail.mfnboer.SkywalkerApplication;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.util.Log;


public class NetworkUtils {
    private static final String LOGTAG = "NetworkUtils";

    public static int getBandwidthKbps() {
        NetworkCapabilities caps = getNetworkCapabilities();

        if (caps == null) {
            Log.w(LOGTAG, "No network capabilities.");
            return -1;
        }

        int kbps = caps.getLinkDownstreamBandwidthKbps();
        Log.d(LOGTAG, "Bandwidth: " + kbps + " kbps");
        return kbps;
    }

    public static boolean isUnmetered() {
        NetworkCapabilities caps = getNetworkCapabilities();

        if (caps == null) {
            Log.w(LOGTAG, "No network capabilities.");
            return false;
        }

        boolean unmetered = caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED);
        Log.d(LOGTAG, "Unmetered: " + unmetered);
        return unmetered;
    }

    private static NetworkCapabilities getNetworkCapabilities() {
        Context context = SkywalkerApplication.getContext();

        if (context == null) {
            Log.w(LOGTAG, "No context.");
            return null;
        }

        ConnectivityManager connectivityManager = context.getSystemService(ConnectivityManager.class);
        Network network = connectivityManager.getActiveNetwork();

        if (network == null) {
            Log.w(LOGTAG, "No active network.");
            return null;
        }

        NetworkCapabilities caps = connectivityManager.getNetworkCapabilities(network);
        return caps;
    }
}
